src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
+﻿using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Composition;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.Extensions.Logging;
 using OmniSharp.Helpers;
 using OmniSharp.Models.Diagnostics;
-using OmniSharp.Roslyn;
+using OmniSharp.Services;
 
-namespace OmniSharp.Workers.Diagnostics
+namespace OmniSharp.Roslyn.CSharp.Services.Diagnostics
 {
-    [Export, Shared]
+    [Shared]
+    [Export(typeof(CSharpDiagnosticService))]
     public class CSharpDiagnosticService
     {
-        private readonly ILogger _logger;
-        private readonly OmniSharpWorkspace _workspace;
-        private readonly object _lock = new object();
+        private readonly ILogger<CSharpDiagnosticService> _logger;
+        private readonly ConcurrentDictionary<ProjectId, (DateTime modified, Project project, CancellationTokenSource workReadySource)> _workQueue =
```

Comment:
saving Project (the snapshot) in the state can be very expensive unless one believe all those will be freom same snapshot such as Workspace.CurrentSolution.

Reply:
one way to get around is that, save projectId and get project by doing workspace.CurrentSolution.GetProject(projectId) when it actually trying to do the work.